### DISCUSSION TITLE
[Proposal]: Nullable unconstrained generic types

### DISCUSSION THREAD

**Discussion by MrTA2** [dotnet/csharplang#4080] [General]
# Nullable unconstrained generic types

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: Implement nullable unconstrained generic types, like `class Foo<T> { T? Bar() { ... } }`

## Motivation
[motivation]: This would get rid of the CS8627, and practically, allow the creation of methods such as FirstOrNull on generic collections, among other things.

## Detailed design
[design]: 
A new type, `UniversalNullable<T>`, is added to the framework:
```
public struct UniversalNullable<T>
{
  public UniversalNullable(T value) { this._hasValue = true; this._value = value; }

  public bool HasValue=>this._hasValue;
  public T Value => this._hasValue ? this._value : throw new Exception("...");

  bool _hasValue; T _value;
}
```
Whenever the C# compiler encounters any usage of `T?`, and T isn't constrained with `class`, `struct`, or a base type, it substitutes the above `UniversalNullable<T>` type. All "get" operations are wrapped in a `.Value` property get (optimizer to inline the getter body); "set" operations are replaced with either `new UniversalNullable<T>()` or `new UniversalNullable<T>(value)`, depending on whether the value is null. All this can be mostly copied from the `Nullable<T>` implementation for structs. Then we can write code like this:

```
class Foo<T> where T : notnull // notnull is needed, otherwise we'll create nullable of nullable which is pointless
{  
  public Foo(T t)=>this._t=t;

  readonly T _t;

  public T? Bar(T? input)
  {
    if(input != null) // compiles to "if(input._hasValue)" - use the field for performance
    {
      // input is not null here
      T res = this.Bar(input); // compiler implicitly resolves "T" by using "._value" (the field), to make it faster: T res = this.Bar(input._value);
      return res; // compiles to "return new UniversalNullable<T>(res);"
    }
    T res2 = this.Bar(input); // warning, possible null; still compile using ".Value" (the property), getter body inlined for performance
    string str = input.ToString(); // warning, possible null; still compile using ".Value.ToString()" - unlike Nullable<T>, never call UniversalNullable<T>.ToString
    string? str2 = input?.ToString(); // OK; if T has interface constraints which have ToString overloads, those are available here (same as above line - even if the above line generates a warning and crashes at runtime)
    return null; // compiles to "return new UniversalNullable<T>();"
  }
  public T Bar(T input) // overloading works
  {
    return input;
  }
  public async Task<T?> BarAsync()
  {
    await Task.Delay(100);
    if(DateTime.Now.Hour < 12) return this.Bar(null); else return this.Bar(this._t);
  }
}
class Extensions
{
  public static T? FirstOrNull<T>(this IEnumerable<T> This)
  {
    foreach(var x in This) return x; //  "return new UniversalNullable<T>(x);"
    return null; // "return new UniversalNullable<T>();"
  }
}
var foo = new Foo<int>(123);
int? x = foo.Bar(null); // resolves Bar(T?); in IL, convert the result from UniversalNullable<int> to Nullable<int> before assignment
int x2 = foo.Bar(456); // resolves Bar(T)
var foo2 = new Foo<string>("abc");
string? y = foo2.Bar(null); // resolves Bar(T?) aka Bar(UniversalNullable<string>); in IL, extract the result from UniversalNullable<string> to string before assignment
string y2 = foo2.Bar("def"); // resolves Bar(T) aka Bar(string)
var foo3 = new Foo<int?>(123); // error
```

## Drawbacks
[drawbacks]: This obviously introduces a minor performance reduction for reference types, since it has the field referencing overhead. (For struct types, it's a given, similarly to Nullable<T>.)  

## Alternatives
[alternatives]: Leave things as is, keep generating CS8627. When type safe `FirstOrNull` methods are needed, implement two versions, one for classes and one for structs.

## Unresolved questions
[unresolved]: None

**Comment by AntiPasha** [comment]
I guess, that it was already covered in [#3297](https://github.com/dotnet/csharplang/issues/3297)

**Comment by MrTA2** [comment]
> 
> 
> I guess, that it was already covered in [#3297](https://github.com/dotnet/csharplang/issues/3297)

I saw it, it's a related discussion, but it doesn't solve the issue once and for all, which this proposal does.

**Comment by YairHalberstadt** [comment]
I think there's a lot of edge cases which need to be considered here.

For example:

```csharp
public interface I
{
    string String { get; }
}

#nullable disable
public M1<T>(T t) where T : I => Console.WriteLine(t == null ? "" : t.String);
# nullable enable

public M2<T>(T? t) where T : notnull, I => M1(t); // What happens here? Does this call `M1<UniversalNullable<T>>`?
```

Personally I think there's way too much that can go wrong with these kind of tricks. It just becomes too risky to put into the language.

**Comment by YairHalberstadt** [comment]
> I'd say yes, if nullable is disabled, it just does T, so your code creates a recursive endless loop.

Where's the loop? it's calling M1?

**Comment by MrTA2** [comment]
> ```cs
> public M2<T>(T? t) where T : notnull, I => M1(t); // What happens here? Does this call `M1<UniversalNullable<T>>`?
> ```
> 
(Sorry I misread your code, didn't notice M1 vs M2)
This would fail to compile, because UniversalNullable<T> doesn't implement I.

**Comment by YairHalberstadt** [comment]
> This would fail to compile, because UniversalNullable doesn't implement I.

What about cases where it's more subtle.

For example, if `T` implements `IEquatable<T>`, and you create a `Dictionary<T?, string>` the dictionary will use ReferenceEquality instead of `IEquatable.Equals` since UniversalNullable doesn't implement `IEquatable`

**Comment by MrTA2** [comment]
> For example, if `T` implements `IEquatable<T>`, and you create a `Dictionary<T?, string>` the dictionary will use ReferenceEquality instead of `IEquatable.Equals` since UniversalNullable doesn't implement `IEquatable`

Yes, but arguably `Dictionary<,>` should have `where TKey:notnull` which would prevent this, and the programmer should know not to do this anyway; think about it, purposefully supplying a nullable key type to a dictionary is nonsensical.

**Comment by YairHalberstadt** [comment]
> Yes, but arguably Dictionary<,> should have where TKey:notnull which would prevent this, and the programmer should know not to do this anyway; think about it, purposefully supplying a nullable key type to a dictionary is nonsensical.

As it happens dictionary throws on a `null` key, but I have implementations of `IDictionary` where this is not the case and it's perfectly sensible.

Besides IEquatable is just an example. There are many cases where code dynamically checks for an interface. What if `T` implements `IDisposable`?

**Comment by 333fred** [comment]
Unconstrained `T?` already means something in C# 9. Changing that will be a breaking change. As much as I would like to change `Nullable<T>` and make it generally applicable to all types, we have 20 years of library design not taking this into account, and we heavily discussed doing something like this as part of nullable only to conclude that that history is too high a barrier.

**Comment by MrTA2** [comment]
> As it happens dictionary throws on a `null` key, but I have implementations of `IDictionary` where this is not the case and it's perfectly sensible.

Actually I just checked and both `Dictionary<,>` and `IDictionary<,>` already do `where TKey:notnull`, so I'm not sure what you're talking about.
> 
> Besides IEquatable is just an example. There are many cases where code dynamically checks for an interface. What if `T` implements `IDisposable`?

_The same problems occur with `Nullable<T>` and struct types. If you had `SomeClass<int?>` and that class were to do `T x=...; if(x is IFormattable) { ... }` it would fail the same way. If we accepted this drawback with `Nullable<T>`, there's no reason to also accept it with `UniversalNullable<T>`._
This is wrong as pointed out below (otherwise `string.Format("{0:000}", (int?)123)` wouldn't work)

**Comment by MrTA2** [comment]
> 
> 
> Unconstrained `T?` already means something in C# 9. Changing that will be a breaking change. As much as I would like to change `Nullable<T>` and make it generally applicable to all types, we have 20 years of library design not taking this into account, and we heavily discussed doing something like this as part of nullable only to conclude that that history is too high a barrier.

I'm not proposing to change `Nullable<T>` - that would definitely be a non-starter. I'm specifically proposing creating a new type, `UniversalNullable<T>`, just for this solution. 

I do understand that unconstrained T? was implemented in C# 9, but I think that was a mistake. I wish I could propose this sooner. Is it at all possible to take that implementation out of C# 9 for now and revisit in C#10?

**Comment by YairHalberstadt** [comment]
`Nullable<T>` does all sorts of runtime magic to implement the same interfaces as `T` implements. You would need runtime magic for this as well.

**Comment by MrTA2** [comment]
> 
> 
> `Nullable<T>` does all sorts of runtime magic to implement the same interfaces as `T` implements. You would need runtime magic for this as well.

True, I retract what I said previously. However, I did point out in my proposal that a lot of these under-the-hood shenanigans can be copied and pasted from `Nullable<T>` or just expanded to accept both `Nullable<T>` and `UniversalNullable<T>`.

**Comment by 333fred** [comment]
> I wish I could propose this sooner.

Again, we considered this. In depth. And we went forward with the conscious decision that we were blocking this feature.

**Comment by YairHalberstadt** [comment]
If you're requiring runtime magic, why not just get this to work as is without introducing a new magic type? The runtime jits separate code for struct `T` and reference type `T` so these would probably be significantly simpler.

**Comment by MrTA2** [comment]
> 
> 
> If you're requiring runtime magic, why not just get this to work as is without introducing a new magic type? The runtime jits separate code for struct `T` and reference type `T` so these would probably be significantly simpler.

The magic type is needed to: 1) house the hasValue and the value; 2) allow method binding and overload resolution cleanly.

**Comment by YairHalberstadt** [comment]
> The magic type is needed to: 1) house the hasValue and the value; 2) allow method binding and overload resolution cleanly

Why?

Just generate the same IL as you would if T wasn't Nullable, and then the runtime will be responsible for handling the case where T is a struct.

**Comment by MrTA2** [comment]
> 
> 
> > The magic type is needed to: 1) house the hasValue and the value; 2) allow method binding and overload resolution cleanly
> 
> Why?
> 
> Just generate the same IL as you would if T wasn't Nullable, and then the runtime will be responsible for handling the case where T is a struct.

I thought about that before. The problem is that for nullability to work with struct types, you need a "hasValue". And once you accept that premise, you realize that you end up creating 2 versions of the code for each unconstrained nullable generic parameter. C# needs to support multiple generic parameters, so it becomes 2^N versions of a method; before starting any execution, a special IL needs to be output which checks all the generics and jumps to the correct version. All this is very inefficient, both space and performance wise. I understand the upside of being able to reuse Nullable<T> just for structs in this case, but the downsides are much worse. Furthermore, when the execution leaves the generic-aware context, more difficulties arise - how do we explain to the callers of our generic methods that sometimes they have to unbox Nullable<T>, and sometimes not? In short, just packaging the references into a UniversalNullable<T>, while adding a minor overhead, is much simpler: no multiple versions of the code; standard logic for get/set operations; etc. I think this is the only viable approach.

**Comment by 333fred** [comment]
> The problem is that for nullability to work with struct types, you need a "hasValue".

So, leaving off the issue compat for a moment, this is only kinda true. In an ideal world, we'd have the runtime niche-fill this. Structs do need a bool for tracking, yes, but reference types do not. They can be represented by the null reference. And if you make the runtime aware of this, and specialize the implementation, then it's totally viable. This is how rust is able to represent `Option<ref T>` as just 0, and would really be the only version of this I'd be interested in. I'll convert this to a discussion for continued conversation, just remember we did consider this at length.

**Reply by MrTA2** [reply]
> 
> 
> > The problem is that for nullability to work with struct types, you need a "hasValue".
> 
> So, leaving off the issue compat for a moment, this is only kinda true. In an ideal world, we'd have the runtime niche-fill this. Structs do need a bool for tracking, yes, but reference types do not. They can be represented by the null reference. And if you make the runtime aware of this, and specialize the implementation, then it's totally viable. This is how rust is able to represent `Option<ref T>` as just 0, and would really be the only version of this I'd be interested in. I'll convert this to a discussion for continued conversation, just remember we did consider this at length.

I'm sure this is possible, but the runtime magic here would be much worse than in what I'm proposing, no? 
My solution only calls for one new type, and changing the "is/as" logic (as Yair brought up) to include the new type in addition to Nullable<>, which i can imagine would be changing "type == Nullable" to "type == Nullable or type == UniversalNullable" in a few places.
If we were to make the runtime understand a single value as either a Nullable<T> or a reference, then at machine level, you end up having up to 2^N versions of the code executing one way or another, whether a) by outputting different IL, or b) by changing JIT to understand this funny IL in this unconstrained T? context and create different assembly code. At the end of the day, checking hasValue is different from doing `!= null` - hasValue is a field get and then a "BR", whereas the `!= null` is a simple "BR". 
What am I missing here?

And you've brought up compatibility a few times and it's still unclear exactly what my proposal would be incompatible with. (Other than the IMHO incorrect new unconstrained T? workaround in C# 9 which isn't "production" yet.)

**Reply by 333fred** [reply]
> And you've brought up compatibility a few times and it's still unclear exactly what my proposal would be incompatible with. (Other than the IMHO incorrect new unconstrained T? workaround in C# 9 which isn't "production" yet.)

This is exactly what I mean. I understand you don't like it. But C# 9 is shipping in a week, and we considered this scenario when making the decision. Unfortunately, that makes this discussion a purely hypothetical one unless we introduced a new dialect of C#, something we're generally very unwilling to do.

> I'm sure this is possible, but the runtime magic here would be much worse than in what I'm proposing, no?

I don't think so, no. I think this solution would end up being very confusing to users.

>  If we were to make the runtime understand a single value as either a Nullable or a reference, then at machine level, you end up having up to 2^N versions of the code executing one way or another, whether a) by outputting different IL, or b) by changing JIT to understand this funny IL in this unconstrained T? context and create different assembly code. At the end of the day, checking hasValue is different from doing != null - hasValue is a field get and then a "BR", whereas the != null is a simple "BR".

I'm not exactly sure where you're getting `2^N` from. And most of this would be specialized away anyway. The runtime is able to share the implementation of generics specialized to reference types already today. The headers of `M<string>` and `M<object>` will be different (which is what enables you to reference the type parameter), but the actual code in the body is shared between both instanciations. And because the runtime would understand that `Nullable<reference type>` is just `reference type`, it would likely be able to share the implementation between the `M<string?>` and `M<object?>` version as well.

**Reply by MrTA2** [reply]
> 
> 
> > And you've brought up compatibility a few times and it's still unclear exactly what my proposal would be incompatible with. (Other than the IMHO incorrect new unconstrained T? workaround in C# 9 which isn't "production" yet.)
> 
> This is exactly what I mean. I understand you don't like it. But C# 9 is shipping in a week, and we considered this scenario when making the decision. Unfortunately, that makes this discussion a purely hypothetical one unless we introduced a new dialect of C#, something we're generally very unwilling to do.

Now that C# 9 is shipping imminently with the workaround, the real solution - either using my proposal, or in some other way - can be implemented using a different symbol (as was proposed for the workaround) like `T~` or whatever.

> 
> > I'm sure this is possible, but the runtime magic here would be much worse than in what I'm proposing, no?
> 
> I don't think so, no. I think this solution would end up being very confusing to users.

I would disagree. C# developers already know and love `Nullable<T>`. We're replicating the same pattern, for a similar reason, in a very adjacent functional area (to solve nullability for structs).

> 
> > If we were to make the runtime understand a single value as either a Nullable or a reference, then at machine level, you end up having up to 2^N versions of the code executing one way or another, whether a) by outputting different IL, or b) by changing JIT to understand this funny IL in this unconstrained T? context and create different assembly code. At the end of the day, checking hasValue is different from doing != null - hasValue is a field get and then a "BR", whereas the != null is a simple "BR".
> 
> I'm not exactly sure where you're getting `2^N` from. 

Imagine this: `void Foo<T1, T2, T3>(T1? t1, T2? t2, T3? t3) { ... }` 
If you call it like so: `Foo<int,int,int>(1, 2, 3);`, it needs to execute one version of the code; `Foo<int, int, string>(1, 2, "c");` another version; `Foo<int, string, int>(1, "b", 3);` yet another; and so on - total of 8 versions, 2^3 - where 3 is the number of generic arguments which are unconstrained and have "nullable use". In real world, this is unlikely, so perhaps you can say, any method can only use one unconstrained generic parameter as a nullable. However, that adds complexity. Not making this illegal forces us to implement the 2^N versions of the code, which is even more complex. There are no good options here.

> 
> And most of this would be specialized away anyway. The runtime is able to share the implementation of generics specialized to reference types already today. The headers of `M<string>` and `M<object>` will be different (which is what enables you to reference the type parameter), but the actual code in the body is shared between both instanciations. And because the runtime would understand that `Nullable<reference type>` is just `reference type`, it would likely be able to share the implementation between the `M<string?>` and `M<object?>` version as well.

Oh yes, I'm not saying it has to create code for each separate type. But it has to create separate bodies for whether it's class or struct, for each generic parameter (which is used in a nullable reference for a variable/return type/etc.). 

This makes compilation very difficult. The compiler needs to examine the whole body of a method and check to see if any parameters, return type, or locals are of type unconstrained T?. If so, then it goes into this special multiple-body outputting mode. 
This is even more complicated by "yield return" and "async" methods; now not only the body of the method has to be split, but also the state machines for those types of methods. 
And this is before we get into class fields. What do we put in the actual heap to store an unconstrained T?, without using a `UniversalNullable`?

**Reply by 333fred** [reply]
> I would disagree. C# developers already know and love Nullable<T>. We're replicating the same pattern, for a similar reason, in a very adjacent functional area (to solve nullability for structs).

Yes, that's the problem. It's a similar, but not the same, pattern. `UniversalNullable<T>` would have to do all the same things as `Nullable<T>`: it would have to implicitly implement interfaces, allow lifted operators, etc. But... it's a new type for some reason?

>  If you call it like so: Foo<int,int,int>(1, 2, 3);, it needs to execute one version of the code; Foo<int, int, string>(1, 2, "c"); another version; Foo<int, string, int>(1, "b", 3); yet another; and so on - total of 8 versions, 2^3 - where 3 is the number of generic arguments which are unconstrained and have "nullable use".

This is true regardless of whether you introduce a new type or not. It doesn't matter if `M<T>(T? t)` is represented by `M<string>(Nullable<string> t)` or `M<string>(UniversalNullable<string> t)`, you still have type explosion. However, as I mentioned previously, if the runtime is aware of the type, it doesn't actually have to be _implementation_ explosion, because the same method body would work for both of these.

> But it has to create separate bodies for whether it's class or struct, for each generic parameter (which is used in a nullable reference for a variable/return type/etc.).

Nothing about `UniversalNullable<T>` changes this calculus.

>  The compiler needs to examine the whole body of a method and check to see if any parameters, return type, or locals are of type unconstrained T?. If so, then it goes into this special multiple-body outputting mode.

No? There's no need for multiple bodies here. If you use an unconstrained `T?`, it's always `Nullable<T>` (in this hypothetical world). No ifs, ands, or buts. It's just that the runtime would be able to share the implementation of the body among all reference types, because it could know that `Nullable<T>` is actually just implemented by `T`.

> And this is before we get into class fields. What do we put in the actual heap to store an unconstrained T?, without using a UniversalNullable?

The same thing we do for `Nullable<T>` today. Which, actually, is already niche filled. The runtime does not need to put the whole `Nullable<T>` struct on the heap, it just needs to have a reference, and uses the null reference to represent `null`. For example, this code will print `System.Int32`, not `Nullable<System.Int32>`, even though the CIL contains an explicit creation of `Nullable<T>` and then box of that `Nullable`:
```cs
using System;
int? i = 1;
object o = i;
Console.WriteLine(o.GetType());
```

**Reply by MrTA2** [reply]
> Yes, that's the problem. It's a similar, but not the same, pattern. `UniversalNullable<T>` would have to do all the same things as `Nullable<T>`: it would have to implicitly implement interfaces, allow lifted operators, etc. But... it's a new type for some reason?

If you are saying you'd prefer to just re-purpose `Nullable<T>` for it, fantastic! I don't disagree. But you were the one who said changing `Nullable<T>` to drop `struct` constraint is fraught with other risks, not me. Personally, I don't see a reason why we can't just repurpose `Nullable<T>`, we're not restricting things, we're allowing more, so existing code should work. I only proposed creating `UniversalNullable<T>` to eliminate possible regressions, and in my mind, creating a new simple type like that is straightforward.

> This is true regardless of whether you introduce a new type or not. It doesn't matter if `M<T>(T? t)` is represented by `M<string>(Nullable<string> t)` or `M<string>(UniversalNullable<string> t)`, you still have type explosion. However, as I mentioned previously, if the runtime is aware of the type, it doesn't actually have to be _implementation_ explosion, because the same method body would work for both of these.
> No? There's no need for multiple bodies here. If you use an unconstrained `T?`, it's always `Nullable<T>` (in this hypothetical world). No ifs, ands, or buts. It's just that the runtime would be able to share the implementation of the body among all reference types, because it could know that `Nullable<T>` is actually just implemented by `T`.

OK I think there's a misunderstanding here. If we use a special type, like `Nullable<T>` or `UniversalNullable<T>`, for both class and struct types, none of this applies. 

This multi-body generation, IMO, is needed if we were to try avoiding the magic type for class types, and only use the existing `Nullable<T>` with the `where T:struct` constraint for struct types. In that case, expressions like `T? x; ... if(x != null) { ... }` would have to generate 2 IL code sequences: one which just compares the value to null, and another which gets _hasValue and compares that to "false" (same as null in IL obviously).

> > And this is before we get into class fields. What do we put in the actual heap to store an unconstrained T?, without using a UniversalNullable?
> 
> The same thing we do for `Nullable<T>` today. Which, actually, is already niche filled. The runtime does not need to put the whole `Nullable<T>` struct on the heap, it just needs to have a reference, and uses the null reference to represent `null`. For example, this code will print `System.Int32`, not `Nullable<System.Int32>`, even though the CIL contains an explicit creation of `Nullable<T>` and then box of that `Nullable`:
> 
> ```cs
> using System;
> int? i = 1;
> object o = i;
> Console.WriteLine(o.GetType());
> ```

This example isn't relevant to this discussion. 
First, when boxing `int?` to `object`, the runtime checks and if it has a value, just boxes that value, and if not, sets the reference to null, as it should. The reverse is also true, if unboxing from `object` to `int?`, it automatically converts null reference to `new Nullable<T>()`, and boxed int to `new Nullable<T>(intValue)`. Calling `GetType` on the boxed object doesn't mean anything; I wasn't referring to boxing; I was referring to storing it as a field of a class, which is stored on the heap.
Second, even if you had `class Foo { public int? X = 123; }`, which is what I was talking about (notice the field), then `new Foo().X.GetType()` is just `System.Int32`, because of the "type lifting" mechanism, not because the runtime magically resized the class and got rid of the hasValue=true and just stored 123 in its place.
Third, again, I was responding to Yair's question (and what I thought was also your question) about not adding `UniversalNullable` and not using `Nullable` for classes, and instead magically changing the behavior to either pass the actual reference for class types and `Nullable<T>` for struct types. Whereas we can hack things left and right to accomplish this on the stack, within the context of a method, we can't really do that for class fields.

**Reply by MrTA2** [reply]
In short, what I think you're saying now, is that why create a new magic type, let's just expand `Nullable<T>` by allowing it to store class types, and get all the type lifting shenanigans for free, and just regression test everything to make sure dropping struct constraint didn't break anything? If so, YES!!! Let's do it.

**Reply by HaloFour** [reply]
> and just regression test everything

😂

**Reply by MrTA2** [reply]
> 
> 
> > and just regression test everything
> 
> 😂

Did I say something funny?

**Reply by 333fred** [reply]
> In short, what I think you're saying now, is that why create a new magic type, let's just expand Nullable<T> by allowing it to store class types, and get all the type lifting shenanigans for free, and just regression test everything to make sure dropping struct constraint didn't break anything? If so, YES!!! Let's do it.

Well, what I'm saying is that it would be wonderful if we didn't have 20 years of BCL and other libraries design that wouldn't interact well with this system. We considered this in depth. And we rejected it as the ship sailed long ago. Theoretically, we could still decide in the future to make `T~` mean this special thing, or decide to introduce a special language and runtime flag to turn `T?` into `Nullable<T>`, not `default(T)`. But we don't think that it is worth the effort that would be required to accomplish.

**Reply by MrTA2** [reply]
> 
> 
> > In short, what I think you're saying now, is that why create a new magic type, let's just expand Nullable by allowing it to store class types, and get all the type lifting shenanigans for free, and just regression test everything to make sure dropping struct constraint didn't break anything? If so, YES!!! Let's do it.
> 
> Well, what I'm saying is that it would be wonderful if we didn't have 20 years of BCL and other libraries design that wouldn't interact well with this system. We considered this in depth. And we rejected it as the ship sailed long ago. Theoretically, we could still decide in the future to make `T~` mean this special thing, or decide to introduce a special language and runtime flag to turn `T?` into `Nullable<T>`, not `default(T)`. But we don't think that it is worth the effort that would be required to accomplish.

OK, so we can't modify `Nullable<T>` because of existing code base, but we don't want to add `UniversalNullable<T>` (which is very little effort) because it's just like `Nullable<T>`? 
Isn't that a circular catch 22 logic? If `Nullable<T>` cannot be used as is, that means `UniversalNullable<T>` is sufficiently different. If `UniversalNullable<T>` is not sufficiently different, then we should be able to use `Nullable<T>`.
I do agree though that cost wise, we need to prioritize appropriately and by no means this is **the most important** new feature.

**Reply by CyrusNajmabadi** [reply]
> Isn't that a circular catch 22 logic? 

No.  it just means there may not be a solution that is sufficient for all your desires.  In effect, the constraints of the system (no pun intended) have locked us into a very limited path here.

**Reply by 333fred** [reply]
> but we don't want to add UniversalNullable<T> (which is very little effort) because it's just like Nullable<T>

Well, that's the thing right? `UniversalNullable<T>` _is_ just like `Nullable<T>`. Users would expect it to work the same way. That would, by necessity, mean the runtime would have to be aware if it. It would mean that things like lifted operators would need to be implemented, interfaces would have to be implicitly implemented. A bunch of language rules would have be created. Likely implicit conversions between `Nullable<T>` and `UniversalNullable<T>` would have to be created. In short: the level of effort required in making `UniversalNullable<T>` would be approximately equal to the effort in making `Nullable<T>` generalized.

**Reply by MrTA2** [reply]
> the level of effort required in making `UniversalNullable<T>` would be approximately equal to the effort in making `Nullable<T>` generalized.

You've said you guys considered this in depth. Is there any info online on those conversations, with respect to potential calamities of generalizing `Nullable<T>`? Like meeting notes? Obviously I don't know as much about all the nasty details as you guys do, so I would be interested in learning more about it.

**Reply by 333fred** [reply]
There have been a lot of discussions over the past few years, more than I could link to in just a few minutes, so for the full picture you'd have to trawling through the notes (which, full disclosure, are not in the greatest shape from ~mid-2018 and before). A couple of relevant conversations from this year though:
- Our initial view of `T??`, where we acknowledged that `T?` meaning `Nullable<T>` always would likely never happen: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-01-08.md#unconstrained-type-parameter-annotation-t
- The followup when implementation revealed that `T??` is ambiguous: https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-06-17.md#t

**Reply by MrTA2** [reply]
I thought about it some more and am convinced a new type (`UniversalNullable<>`) is called for - generalizing `Nullable<>` would be a mistake. The "hacks" inside the framework around boxing/unboxing `Nullable<>` don't translate perfectly to nullable generic argument when it's a reference type; it's one of those cases where similarity does not mean equality - copying and pasting, then modifying, is the solution here, IMHO. Also, it's just a code smell to reuse a type for a different purpose; when one sees `Nullable<>`, they should know right away it's for value type nullables only. And, of course there is the regression aspect: countless amount of code was written which relies on reflection and looks for `Nullable<>` (whether or not it's a good idea to use these internal hacks) - generalizing that type would be a massive breaking change for the world. Whereas `UniversalNullable<>` is a new type, for a new scenario, without any regressions.

Obviously, it's a bit too late BECAUSE SOMEBODY DECIDED to do the worst possible solution (IMHO). But can we please, please add `UniversalNullable<>` approach with a different symbol? Like `T~` or something? Any symbol or keyword - just make it work properly.

**Reply by CyrusNajmabadi** [reply]
> just make it work properly.

What does that mean here?

**Reply by TahirAhmadov** [reply]
I mean that I think that what C# 9 did with `T?` was a mistake for 2 reasons. 1, it adds an inconsistent behavior between references and values - null vs "zero", whatever that happens to be. 2, it makes it impossible to do a proper solution (like the one I described above) with the syntax which would intuitively fit it - `T?`. 

Water under the bridge. Going forward, and I understand it's not a high priority item (especially given the rather poor IMHO solution of "defaultability" being denoted by `T?`), I would implore the language team to consider doing something like `UniversalNullable<>` with a different symbol or keyword.

**Reply by CyrusNajmabadi** [reply]
> I would implore the language team to consider doing something like UniversalNullable<> with a different symbol or keyword.

I'm not sure such a concept can exist.  This is literally not how the CLR works here.

**Reply by TahirAhmadov** [reply]
Why not? What part of my proposal exactly goes against how the CLR works?

**Reply by CyrusNajmabadi** [reply]
Can you link to your proposal and indicate what it would compile down to.

**Reply by TahirAhmadov** [reply]
Please see thread post. I apologize in advance if I missed anything or the formatting is not up to standard - I don't create these often.

**Comment by TahirAhmadov** [comment]
I swear, I knew this would happen. I ran into a real world problem specifically because of the lack of this feature. Consider the following:
```cs
class Base<T> where T : notnull
{
  public virtual T? Prop { get; set; } // I can't do this, because if T is struct, this becomes just T
  public virtual Null<T> Prop { get; set; } // so I had to write my own "UniversalNullable"
}
class ChildClass<T> : Base<T> where T : class
{
}
class ChildStruct<T> : Base<T> where T : struct
{
}
public struct Null<T> where T : notnull
{
  public static implicit operator Null<T>(T value) => new Null<T>(value);
  public static explicit operator T(Null<T> nullable) => nullable.Value;
  public static bool operator ==(Null<T> first, Null<T> second) => first.Equals(second);
  public static bool operator !=(Null<T> first, Null<T> second) => !first.Equals(second);
  //public static bool operator ==(Null<T> nullable, object? obj) => nullable.Equals(obj);
  //public static bool operator !=(Null<T> nullable, object? obj) => !nullable.Equals(obj);
  //public static bool operator ==(object? obj, Null<T> nullable) => nullable.Equals(obj);
  //public static bool operator !=(object? obj, Null<T> nullable) => !nullable.Equals(obj);
  //public static explicit operator Null<T>(object? obj) => default;

  public static Null<T> Unbox(object? obj) => obj != null ? new Null<T>((T)obj) : default;

  public Null(T value) { this._hasValue = true; this._value = value; }
  public bool HasValue => this._hasValue;
  public T Value { get { if (!this._hasValue) throw new InvalidOperationException("Nullable object must have a value."); return this._value; } }
  public object? Box() => this._hasValue ? this._value : null;
  public override bool Equals(object? obj) { ... }
  public bool Equals(Null<T> obj) { ... }
  public override int GetHashCode() { ... }
  public override string ToString() { if (!this._hasValue) return ""; return this._value.ToString(); }
  readonly bool _hasValue;
  readonly T _value;
}
public static class Null
{
  public static Null<T> Create<T>(T value) where T : notnull => new Null<T>(value);
  public static Null<T> CreateClass<T>(T? value) where T : class => value != null ? new Null<T>(value) : default;
  public static Null<T> Create<T>(T? nullable) where T : struct => nullable.HasValue ? new Null<T>(nullable.Value) : default;
}
public static class _Null_T_Extensions
{
  public static T? GetOrNull<T>(this Null<T> This) where T : class => This.HasValue ? This.Value : null;
  public static T? GetOrNullStruct<T>(this Null<T> This) where T : struct => This.HasValue ? This.Value : null;
  public static TProp? IfNotNull<TValue, TProp>(this Null<TValue> This, Func<TValue, TProp> getProp) where TValue : notnull where TProp : class
	{ if (This != default) return getProp(This.Value); return null; }
  public static TProp? IfNotNullStruct<TValue, TProp>(this Null<TValue> This, Func<TValue, TProp> getProp) where TValue : notnull where TProp : struct
  { if (This != default) return getProp(This.Value); return null; }
}
```
Basically, almost everything C# and .NET do today behind the scenes for `Nullable<T>`, in terms of boxing and unboxing, equality, and conversions, I had to write myself and call it explicitly. That's the first big problem - I have to remember how to code "the right way" to use my `Null<T>`. Second, the syntax is off; I couldn't make it work so that `Null<T> x; if(x != null) { }` and `x = null;` all work - the only consistent syntax was `if(x == default)` and `x = default;`; so in code using this, `default` means `null`. Third, there were issues because interfaces don't work with implicit operators, so I had to add static methods to handle those. Fourth, interface propagation and type lifting don't work because that would require CLR changes (thankfully, I don't need them yet). Fifth, operators like `?.` and `??` don't work, so I had to create extension methods `IfNotNull`, etc.
Needless to say, none of these would be an issue if this feature was added to the language.
